### PR TITLE
FOGL-2625 system REST API tests added for statistics and statistics history

### PR DESCRIPTION
**Pull request by ashish-jabble**
Checked on
- [x] ub16.04
- [x] raspbian

```
$ pytest -s -vv test_statistics.py
============================================================= test session starts =============================================================
platform linux -- Python 3.5.3, pytest-3.6.0, py-1.8.0, pluggy-0.6.0 -- /usr/local/bin/python3.5
cachedir: ../.pytest_cache
rootdir: /home/foglamp/Development/systemTests/FogLAMP/tests/system/python, inifile: pytest.ini
plugins: mock-1.10.0, cov-2.5.1, asyncio-0.8.0, allure-adaptor-1.7.10, aiohttp-0.3.0
collected 10 items                                                                                                                            

../TestStatistics test_cleanup <- api/test_statistics.py FogLAMP killed.
This script will remove all data stored in the SQLite3 datafile '/home/foglamp/Development/systemTests/FogLAMP/data/foglamp.db'
Enter YES if you want to continue: Starting FogLAMP v1.5.1......
FogLAMP started.
PASSED
../TestStatistics test_default_statistics <- api/test_statistics.py PASSED
../TestStatistics test_default_statistics_history <- api/test_statistics.py PASSED
../TestStatistics test_statistics_history_with_stats_collector_schedule <- api/test_statistics.py PASSED
../TestStatistics test_statistics_history_with_params <- api/test_statistics.py PASSED
../TestStatistics test_statistics_history_with_params <- api/test_statistics.py PASSED
../TestStatistics test_statistics_history_with_params <- api/test_statistics.py PASSED
../TestStatistics test_statistics_history_with_params <- api/test_statistics.py PASSED
../TestStatistics test_statistics_history_with_params <- api/test_statistics.py PASSED
../TestStatistics test_statistics_history_with_service_enabled <- api/test_statistics.py Cloning into '/tmp/foglamp-south-coap'...
remote: Enumerating objects: 3, done.
remote: Counting objects: 100% (3/3), done.
remote: Compressing objects: 100% (3/3), done.
remote: Total 262 (delta 0), reused 0 (delta 0), pack-reused 259
Receiving objects: 100% (262/262), 44.19 KiB | 0 bytes/s, done.
Resolving deltas: 100% (88/88), done.
Checking connectivity... done.
Collecting aiocoap==0.3 (from -r /tmp/foglamp-south-coap/python/requirements-coap.txt (line 1))
Collecting cbor2==4.0.0 (from -r /tmp/foglamp-south-coap/python/requirements-coap.txt (line 2))
Installing collected packages: aiocoap, cbor2
  The scripts aiocoap-client and aiocoap-proxy are installed in '/home/foglamp/.local/bin' which is not on PATH.
  Consider adding this directory to PATH or, if you prefer to suppress this warning, use --no-warn-script-location.
Successfully installed aiocoap-0.3 cbor2-4.0.0
coap plugin is installed.
>>> Make sure south CoAP plugin service is running 
 & listening on specified host and port 

Total Statistics:

Start Time: 2019-04-01 17:27:23.146078
End Time:   2019-04-01 17:27:23.158111

Total Messages Transferred: 1
Total Bytes Transferred:    288

Total Iterations: 1
Total Messages per Iteration: 1.0
Total Bytes per Iteration:    288.0

Min messages/second: 83.10479514667996
Max messages/second: 83.10479514667996
Avg messages/second: 83.10479514667996

Min Bytes/second: 23934.18100224383
Max Bytes/second: 23934.18100224383
Avg Bytes/second: 23934.18100224383
/home/foglamp/Development/systemTests/FogLAMP/tests/system/python/api
PASSED

====== 10 passed in 66.17 seconds =====
```

### CODE REVIEW

File: tests/system/python/api/test_statistics.py [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+from collections import Counter
+import pytest
+
+__author__ = "Ashish Jabble"
+__copyright__ = "Copyright (c) 2019 Dianomic Systems"
+__license__ = "Apache 2.0"
+__version__ = "${VERSION}"
+
+
+KEYS = {'DISCARDED', 'PURGED', 'history_ts', 'READINGS', 'UNSENT', 'UNSNPURGED', 'BUFFERED'}
```
Comment (praveen-garg) [inline tests/system/python/api/test_statistics.py]: history_ts is not a stats key, and better to name this list as STATS_KEYS

File: tests/system/python/api/test_statistics.py [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+        assert 200 == r.status
+        r = r.read().decode()
+        jdoc = json.loads(r)
+        assert len(jdoc), "No data found"
+        assert 15 == jdoc['interval']
+        assert {} == jdoc['statistics'][0]
+
+    def test_statistics_history_with_stats_collector_schedule(self, foglamp_url, wait_time):
+        # wait for sometime for stats collector schedule to start
+        time.sleep(wait_time * 3)
```
Comment (praveen-garg) [inline tests/system/python/api/test_statistics.py]: let's not keep waiting for 15 seconds, but run stats collector every 2 -5 seconds for test

File: tests/system/python/api/test_statistics.py [hunk 3]
```diff
@@ -0,0 +1,98 @@
...
+        conn.request("GET", '/foglamp/statistics/history')
+        r = conn.getresponse()
+        assert 200 == r.status
+        r = r.read().decode()
+        jdoc = json.loads(r)
+        assert len(jdoc), "No data found"
+        assert 15 == jdoc['interval']
+        assert 1 == len(jdoc['statistics'])
+        statistics = jdoc['statistics'][0]
+        assert Counter(KEYS) == Counter(statistics.keys())
```
Comment (praveen-garg) [inline tests/system/python/api/test_statistics.py]: assert Counter(KEYS + ['history_ts'] ) == Counter(stats_history.keys())

File: tests/system/python/api/test_statistics.py [hunk 4]
```diff
@@ -0,0 +1,98 @@
...
+        r = conn.getresponse()
+        assert 200 == r.status
+        r = r.read().decode()
+        jdoc = json.loads(r)
+        assert len(jdoc), "No data found"
+        assert 15 == jdoc['interval']
+        assert 1 == len(jdoc['statistics'])
+        assert Counter(keys) == Counter(jdoc['statistics'][0].keys())
+
+    @pytest.mark.skip(reason="TODO: will add in future; as a bit tricky one")
```
Comment (praveen-garg) [inline tests/system/python/api/test_statistics.py]: better to remove it, as we may never do it as a system test (with real data). 
please check that its covered with unit test.

File: tests/system/python/api/test_statistics.py (hunk 2 above)
Comment (praveen-garg) [reply tests/system/python/api/test_statistics.py]: And  with south plugin is enabled before, check stats and history are collected for that. (as the last test)

Comment (ashish-jabble) [reply tests/system/python/api/test_statistics.py]: Its not the responsibility of System REST API tests; as these scenarios are covered in `e2e` tests

Comment (praveen-garg) [reply tests/system/python/api/test_statistics.py]: for stats yes, but not the stats history.

File: tests/system/python/api/test_statistics.py (hunk 4 above)
Comment (ashish-jabble) [reply tests/system/python/api/test_statistics.py]: Done

File: tests/system/python/api/test_statistics.py (hunk 2 above)
Comment (ashish-jabble) [reply tests/system/python/api/test_statistics.py]: Added

File: tests/system/python/api/test_statistics.py [hunk 5]
```diff
@@ -0,0 +1,179 @@
...
+
+    def test_default_statistics(self, foglamp_url):
+        conn = http.client.HTTPConnection(foglamp_url)
+        conn.request("GET", '/foglamp/statistics')
+        r = conn.getresponse()
+        assert 200 == r.status
+        r = r.read().decode()
+        jdoc = json.loads(r)
+        assert len(jdoc), "No data found"
+        assert 6 == len(jdoc)
```
Comment (praveen-garg) [inline tests/system/python/api/test_statistics.py]: do you mean assert len(STATS_KEYS) == len(jdoc)

File: tests/system/python/api/test_statistics.py (hunk 2 above)
Comment (praveen-garg) [reply tests/system/python/api/test_statistics.py]: def test_statistics_history_with_stats_collector_schedule(self, foglamp_url, wait_time, retries):
          time.sleep(wait_time * retries)

File: tests/system/python/api/test_statistics.py [hunk 6]
```diff
@@ -0,0 +1,179 @@
...
+        assert 200 == r.status
+        r = r.read().decode()
+        jdoc = json.loads(r)
+        assert len(jdoc), "No data found"
+        assert 15 == jdoc['interval']
+        assert 1 == len(jdoc['statistics'])
+        assert Counter(keys) == Counter(jdoc['statistics'][0].keys())
+
+    def test_statistics_history_with_service_enabled(self, start_south_coap, foglamp_url, wait_time):
+        time.sleep(wait_time)
```
Comment (praveen-garg) [inline tests/system/python/api/test_statistics.py]: why this sleep is needed?

File: tests/system/python/api/test_statistics.py [hunk 7]
```diff
@@ -0,0 +1,179 @@
...
+        conn.request("GET", '/foglamp/statistics')
+        r = conn.getresponse()
+        assert 200 == r.status
+        r = r.read().decode()
+        jdoc = json.loads(r)
+        assert len(jdoc), "No data found"
+        stats = utils.serialize_stats_map(jdoc)
+        assert 1 == stats[ASSET_NAME.upper()]
+        assert 1 == stats['READINGS']
+        time.sleep(wait_time * 2)
```
Comment (praveen-garg) [inline tests/system/python/api/test_statistics.py]: use retries  or (max) retries -1

File: tests/system/python/api/test_statistics.py [hunk 8]
```diff
@@ -0,0 +1,184 @@
...
+        assert Counter(STATS_HISTORY_KEYS) == Counter(statistics.keys())
+        del statistics['history_ts']
+        assert Counter([0, 0, 0, 0, 0, 0]) == Counter(statistics.values())
+
+    @pytest.mark.parametrize("request_params, keys", [
+        ('', STATS_HISTORY_KEYS),
+        ('?limit=1', STATS_HISTORY_KEYS),
+        ('?key=READINGS', {'history_ts', 'READINGS'}),
+        ('?key=READINGS&limit=1', {'history_ts', 'READINGS'}),
+        ('?key=READINGS&limit=0', {}),
```
Comment (Singhal-Vaibhav) [inline tests/system/python/api/test_statistics.py]: Add test for time based queries - minutes,hours,days

Comment (ashish-jabble) [reply tests/system/python/api/test_statistics.py]: https://github.com/foglamp/FogLAMP/pull/1483#discussion_r269464887

**Pull request merged**